### ISSUE TITLE
Large plotly plots will crash streamsync, while dash has no issues showing the same...

### ISSUE THREAD

**Issue by thondeboer** [streamsync-cloud/streamsync#190]
I am really trying to avoid using dash, since I think streamsync is so much easier to work with, but I want to use a lot of complicated plotly plots, such as heatmaps showing 100,000 x 100 data and streamsync has a problem with these complicated plots, while DASH has no issues and is pretty fast.

When I try anything above 8,000 x 100 (rows and cols), streamsync will typically spin for minutes and often just crashes or make firefox crash. Dash has no issues, showing the heatmap up to 100,000 rows x 100 cols in a matter of seconds...

I suspect the plotly rendering code in streamsync is at fault here, but I have not yet looked at it myself, but hopefully one of you can or point me in the right direction to see how we can address this...I really have to go back to dash, but I need my app to work with large data...

Here's my toy examples in `streamsync`
```python
import streamsync as ss
import plotly.graph_objects as go
import numpy as np

def _create_heatmap(rows:int = 1000, cols:int = 100) -> go.Figure:
    return go.Figure(
        data=go.Heatmap(
                z=np.random.rand(int(rows), int(cols)),
            ),
        layout=go.Layout(
            title=f'Heatmap {rows} x {cols}',
            height=1200
        )
    )

def update_heatmap(state = None, payload = None, context = None):
    """Update heatmap."""
    state['heatmap'] =  _create_heatmap(
        state['rows'],
        state['cols']
    )
 
initial_state = ss.init_state({
    "heatmap": None,
    "rows": 1000,
    "cols": 100
})

update_heatmap(state=initial_state)
```

`UI`
```json
{
    "metadata": {
        "streamsync_version": "0.2.8"
    },
    "components": {
        "root": {
            "id": "root",
            "type": "root",
            "content": {
                "appName": "My App"
            },
            "parentId": null,
            "position": 0,
            "handlers": {},
            "visible": true
        },
        "k03d20wgwa3b4p6c": {
            "id": "k03d20wgwa3b4p6c",
            "type": "page",
            "content": {},
            "parentId": "root",
            "position": 0,
            "handlers": {},
            "visible": true
        },
        "fj4o4rgrkiiouk3u": {
            "id": "fj4o4rgrkiiouk3u",
            "type": "plotlygraph",
            "content": {
                "spec": "@{heatmap}"
            },
            "parentId": "k03d20wgwa3b4p6c",
            "position": 1,
            "handlers": {},
            "visible": true
        },
        "uisg1xnhanmdoijo": {
            "id": "uisg1xnhanmdoijo",
            "type": "numberinput",
            "parentId": "k03d20wgwa3b4p6c",
            "content": {
                "label": "Number of rows",
                "minValue": "1000",
                "maxValue": "100000",
                "valueStep": "1000"
            },
            "handlers": {
                "ss-number-change-finish": "update_heatmap"
            },
            "position": 0,
            "visible": true,
            "binding": {
                "eventType": "ss-number-change",
                "stateRef": "rows"
            }
        }
    }
}
```

Also need to ensure streamsync does not time out, so had to add a `serve.py` file as well to set the ws_ping_timeout

```python
import uvicorn
import streamsync.serve
from fastapi import FastAPI, Response
import os

mode = "edit"
app_dir = os.path.dirname(__file__)

sub_asgi_app_1 = streamsync.serve.get_asgi_app(app_dir, mode, enable_remote_edit=True)
root_asgi_app = FastAPI(lifespan=sub_asgi_app_1.router.lifespan_context)

root_asgi_app.mount("/", sub_asgi_app_1)

uvicorn.run(root_asgi_app,
    host="localhost",
    port=3006,
    log_level="warning",
    ws_max_size=streamsync.serve.MAX_WEBSOCKET_MESSAGE_SIZE, ws_ping_timeout=6000)
```

Here's the equivalent DASH program

```python
import dash
from dash import dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np

# Initialize the Dash app
app = dash.Dash(__name__)

# Define the app layout
app.layout = html.Div([
    dcc.Input(
        id='row-input',
        type='number',
        min=1000,
        max=100000,
        step=1000,
        value=1000,  # Default value
        style={'marginRight':'10px'}
    ),
    html.Button('Update Heatmap', id='update-button'),
    dcc.Graph(id='heatmap-graph')
])

def _create_heatmap(rows:int = 1000, cols:int = 100) -> go.Figure:
    return go.Figure(
        data=go.Heatmap(
                z=np.random.rand(int(rows), int(cols)),
            ),
        layout=go.Layout(
            title=f'Heatmap {rows} x {cols}',
            height=1200
        )
    )

# Define callback to update heatmap
@app.callback(
    Output('heatmap-graph', 'figure'),
    [Input('update-button', 'n_clicks')],
    [dash.dependencies.State('row-input', 'value')]
)
def update_heatmap(n_clicks, rows):
    if rows is None:
        raise dash.exceptions.PreventUpdate
    return _create_heatmap(rows, 100)

# Run the app
if __name__ == '__main__':
    app.run_server(debug=True)
```

**Comment by thondeboer** [comment]
Here's proof of DASH plot
![image](https://github.com/streamsync-cloud/streamsync/assets/3422829/fdba98f1-2b47-441e-893d-c2364572078d)

and STREAMSYNC (can do 5000 but is already struggling (Showing green bar for about 30 secods)
![image](https://github.com/streamsync-cloud/streamsync/assets/3422829/f32c0220-d4ef-47bc-b736-c83c84378d39)

**Comment by FabienArcellier** [comment]
Thanks a lot providing example. It help me a lot to confirm the problem root cause.

The problem is due to the json serializer which consumes too many resources for plotly objects. A mechanism is already in place but is not used for plotly objects. It is easy to implement. I will try to open a PR today or tomorrow.

You need to modify the ``core.py:StateSerializer`` class by adding `plotly.graph_objs._figure.Figure` support.

Here is an example of streamsync with the change. 100000 is working but does not display the heatmap on my computer, neither on streamsync or dash.

![Peek 2023-12-26 05-31](https://github.com/streamsync-cloud/streamsync/assets/159559/6cf843bd-19a6-4da7-a8fe-96db6e226c8b)


```
class StateSerialiser:

    """
    Serialises user state values before sending them to the front end.
    Provides JSON-compatible values, including data URLs for binary data.
    """

    def serialise(self, v: Any) -> Union[Dict, List, str, bool, int, float, None]:
        if isinstance(v, StateProxy):
            return self._serialise_dict_recursively(v.to_dict())
        if isinstance(v, (FileWrapper, BytesWrapper)):
            return self._serialise_ss_wrapper(v)
        if isinstance(v, (datetime.datetime, datetime.date)):
            return str(v)
        if isinstance(v, bytes):
            return self.serialise(BytesWrapper(v))
        if isinstance(v, dict):
            return self._serialise_dict_recursively(v)
        if isinstance(v, list):
            return self._serialise_list_recursively(v)
        if isinstance(v, (str, bool)):
            return v
        if v is None:
            return v

        # Checking the MRO allows to determine object type without creating dependencies
        # to these packages

        v_mro = [
            f"{x.__module__}.{x.__name__}" for x in inspect.getmro(type(v))]

        if isinstance(v, (int, float)):
            if "numpy.float64" in v_mro:
                return float(v)
            if math.isnan(v):
                return None
            return v

        if "matplotlib.figure.Figure" in v_mro:
            return self._serialise_matplotlib_fig(v)
        if "plotly.graph_objs._figure.Figure" in v_mro:
            return v.to_json()
        if "numpy.float64" in v_mro:
            return float(v)
        if "numpy.ndarray" in v_mro:
            return self._serialise_list_recursively(v.tolist())
        if "pandas.core.frame.DataFrame" in v_mro:
            return self._serialise_pandas_dataframe(v)
        if "pyarrow.lib.Table" in v_mro:
            return self._serialise_pyarrow_table(v)

        if hasattr(v, "to_dict") and callable(v.to_dict):
            # Covers Altair charts, Plotly graphs
            return self._serialise_dict_recursively(v.to_dict())

        raise StateSerialiserException(
            f"Object of type { type(v) } (MRO: {v_mro}) cannot be serialised.")
```

**Comment by thondeboer** [comment]
Absolutely brilliant! Thanks for this..it's a life saver...Easiest implementation even :)

**Comment by ramedina86** [comment]
Published in v0.3.0, thanks @thondeboer for this and for being a friend of the project!
